### PR TITLE
Remove callable calls from loop termination condition, project wide

### PR DESCRIPTION
**Pull request by frankmayer**
### Summary of Changes
- Remove callable calls from loop termination condition. Will speed up those loops.

Note. There has been no `foreach` conversion on purpose, This might take place for some of those, in another PR.

### Testing Instructions

None, should not change behavior

### Documentation Changes Required

None.

### CODE REVIEW

File: libraries/joomla/database/driver/postgresql.php [hunk 1]
```diff
@@ -1346,7 +1346,7 @@ public function replacePrefix($query, $prefix = '#__')
 
 			$explodedQuery = explode('\'', $query);
 
-			for ($nIndex = 0; $nIndex < count($explodedQuery); $nIndex = $nIndex + 2)
+			for ($nIndex = 0, $nIndexMax = count($explodedQuery); $nIndex < $nIndexMax; $nIndex+= 2)
```
Comment (andrepereiradasilva) [inline libraries/joomla/database/driver/postgresql.php]: Cs lacks one space in nindex sum

File: libraries/joomla/database/driver/postgresql.php [hunk 2]
```diff
@@ -1336,7 +1336,7 @@ public function replacePrefix($query, $prefix = '#__')
 			{
 				$query = explode('setval', $query);
 
-				for ($nIndex = 1; $nIndex < count($query); $nIndex = $nIndex + 2)
+				for ($nIndex = 1, $nIndexMax = count($query); $nIndex < $nIndexMax; $nIndex+= 2)
```
Comment (andrepereiradasilva) [inline libraries/joomla/database/driver/postgresql.php]: Cs lacks one space in nindex sum

File: libraries/joomla/database/driver/postgresql.php [hunk 3]
```diff
@@ -1323,7 +1323,7 @@ public function replacePrefix($query, $prefix = '#__')
 			{
 				$query = explode('nextval', $query);
 
-				for ($nIndex = 1; $nIndex < count($query); $nIndex = $nIndex + 2)
+				for ($nIndex = 1, $nIndexMax = count($query); $nIndex < $nIndexMax; $nIndex+= 2)
```
Comment (andrepereiradasilva) [inline libraries/joomla/database/driver/postgresql.php]: Cs lacks one space in nindex sum

File: libraries/joomla/database/driver/postgresql.php [hunk 4]
```diff
@@ -1310,7 +1310,7 @@ public function replacePrefix($query, $prefix = '#__')
 			{
 				$query = explode('currval', $query);
 
-				for ($nIndex = 1; $nIndex < count($query); $nIndex = $nIndex + 2)
+				for ($nIndex = 1, $nIndexMax = count($query); $nIndex < $nIndexMax; $nIndex+= 2)
```
Comment (andrepereiradasilva) [inline libraries/joomla/database/driver/postgresql.php]: Cs lacks one space in nindex sum

File: administrator/components/com_joomlaupdate/restore.php [hunk 5]
```diff
@@ -7377,7 +7377,7 @@ public static function AESDecryptCtr($ciphertext, $password, $nBits)
 
 			$plaintxtByte = array();
 
-			for ($i = 0; $i < strlen($ciphertext[$b]); $i++)
+			for ($i = 0, $iMax = strlen($ciphertext[$b]); $i < $iMax; $i++)
 			{
```
Comment (andrepereiradasilva) [inline administrator/components/com_joomlaupdate/restore.php]: As said restore is 3 party. Please remove this change

**Review by andrepereiradasilva — APPROVED** [review]

**Comment by anibalsanchez** [comment]
I have tested this item :white_check_mark: successfully on 388ab4ff49bfcf16f19ab17a456395c6a0b2b561<br /><br />Code review OK<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/13258">issues.joomla.org/tracker/joomla-cms/13258</a>.</sub>

File: administrator/templates/hathor/html/com_templates/template/default_tree.php [hunk 6]
```diff
@@ -21,7 +21,7 @@
 
 			if (count($fileArray) >= count($keyArray))
 			{
-				for ($i = 0; $i < count($keyArray); $i++)
+				for ($i = 0, $iMax = count($keyArray); $i < $iMax; $i++)
```
Comment (andrepereiradasilva) [inline administrator/templates/hathor/html/com_templates/template/default_tree.php]: ok, but why still do the `count($keyArray);` twice? (in if and in for)

File: administrator/components/com_templates/views/template/tmpl/default_tree.php [hunk 7]
```diff
@@ -21,7 +21,7 @@
 
 			if (count($fileArray) >= count($keyArray))
 			{
-				for ($i = 0; $i < count($keyArray); $i++)
+				for ($i = 0, $iMax = count($keyArray); $i < $iMax; $i++)
```
Comment (andrepereiradasilva) [inline administrator/components/com_templates/views/template/tmpl/default_tree.php]: ok, but why still do the `count($keyArray);` twice? (in if and in for)

**Review by andrepereiradasilva — COMMENTED** [review]
just some questins, the rest seems fine

File: administrator/templates/hathor/html/com_templates/template/default_tree.php (hunk 6 above)
Comment (frankmayer) [inline administrator/templates/hathor/html/com_templates/template/default_tree.php]: Nice catch!! Changed. Will be up with next commit.

File: administrator/components/com_templates/views/template/tmpl/default_tree.php (hunk 7 above)
Comment (frankmayer) [inline administrator/components/com_templates/views/template/tmpl/default_tree.php]: As above... 
Plus, it was actually counting the $keyArray three times. Also did a strict comparison of integer values.

File: administrator/components/com_templates/views/template/tmpl/default_tree.php [hunk 8]
```diff
@@ -19,17 +19,19 @@
 			$fileArray = explode('/', $this->fileName);
 			$count     = 0;
 
-			if (count($fileArray) >= count($keyArray))
+			$keyArrayCount = count($keyArray);
+
+			if (count($fileArray) >= $keyArrayCount)
 			{
-				for ($i = 0; $i < count($keyArray); $i++)
+				for ($i = 0, $iMax = $keyArrayCount; $i < $iMax; $i++)
```
Comment (Quy) [inline administrator/components/com_templates/views/template/tmpl/default_tree.php]: Don't need this variable `$iMax`.

File: administrator/templates/hathor/html/com_templates/template/default_tree.php [hunk 9]
```diff
@@ -19,17 +19,19 @@
 			$fileArray = explode('/', $this->fileName);
 			$count     = 0;
 
-			if (count($fileArray) >= count($keyArray))
+			$keyArrayCount = count($keyArray);
+
+			if (count($fileArray) >= $keyArrayCount)
 			{
-				for ($i = 0; $i < count($keyArray); $i++)
+				for ($i = 0, $iMax = $keyArrayCount; $i < $iMax; $i++)
```
Comment (Quy) [inline administrator/templates/hathor/html/com_templates/template/default_tree.php]: Don't need this variable `$iMax`.

File: administrator/components/com_templates/views/template/tmpl/default_tree.php (hunk 8 above)
Comment (frankmayer) [inline administrator/components/com_templates/views/template/tmpl/default_tree.php]: Indeed!! Long day...

**Comment by andrepereiradasilva** [comment]
I have tested this item :white_check_mark: successfully on 2ff0334ecd5afb6daddb80370856f6c7f57e724c<br /><br />On code review<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/13258">issues.joomla.org/tracker/joomla-cms/13258</a>.</sub>

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 2ff0334ecd5afb6daddb80370856f6c7f57e724c<br /><br />Code review.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/13258">issues.joomla.org/tracker/joomla-cms/13258</a>.</sub>

**Comment by franz-wohlkoenig** [comment]
RTC after two successful tests.

**Comment by wilsonge** [comment]
@frankmayer would you be willing to do this to the framework database package here https://github.com/joomla-framework/database so it's included in the 4.0 branch please

**Comment by frankmayer** [comment]
@wilsonge yes, I can do that, though not at the moment. Got a lot on my plate these days ;)
Will make a note and revisit after a few days.

**Comment by frankmayer** [comment]
@wilsonge done: https://github.com/joomla-framework/database/pull/95 and https://github.com/joomla-framework/database/pull/96
Don't know if both are needed. I saw what seems to be the 1.x line and also the 2.x-dev, so I thought I'd do the PR against both...

**Pull request merged**